text/0000-consensus-api.md

```diff
@@ -0,0 +1,320 @@
+| --- | --- |
+| CommitBlock(block_id) | Set the chain head to the given block id. The Chain Controller would handle the details here. |
+| DropBlock(block_id) | Remove the given block from the system. The main purpose of this method is to allow the Consensus Engine to notify the rest of the system that a block is invalid from the perspective of consensus. |
+
+### Query Methods
+The following methods will be provided to Consensus Engines for getting
+information from the validator.
+
+| Method | Description |
+| --- | --- |
+| GetSetting(setting) -> data | Read the current value of the setting. Supports R4. |
+| GetState(address) -> data | This is needed to read values from arbitrary addresses used by consensus (eg., validator registry). Supports R5. |
+| GetBlock(block_id) -> block | Retrieve consensus-related information about a block. |
+
+While the consensus engine should maintain a cache of blocks internally, this method is provided to:
+1. Allow cache entries to expire without losing the block forever
+2. Allow the engine to handle getting a block before its parent
+
+### Engine Methods
+The following methods will be implemented by Consensus Engines. They are used
+to initialize and shutdown the engine, when a new validator starts up, or when
+the consensus engine changes.
+
+| Method | Description |
+| --- | --- |
+| Startup() | Startup and synchronize with any peers if necessary. |
+| Shutdown() | Shutdown and notify peers if necessary. |
+
+## Consensus Engine Architecture Integration
+The following describes the interactions between the consensus engine and the
+other major architectural components of the validator.
+
+![consensus engine architecture](../images/consensus-engine-architecture.svg "Consensus Engine Architecture")
+
+**Consensus Engine**
+
+The Consensus Engine is a separate process that implements a consensus
+algorithm and communicates with the Consensus Proxy through the Dispatcher. It
+can also communicate with peers using consensus messages. Its role is to direct
+block creation and chain management. It must implement the interface described
+above.
+
+**Consensus Proxy**
+
+The Consensus Proxy mediates interactions between the Consensus Engine and the
+rest of the Validator. This includes marshaling and unmarshaling messages from
+the Dispatcher, passing block notifications to the Consensus Engine, and
+passing commands from the Consensus Engine to the Chain Controller and Block
+Publisher.
+
+**Block Validator**
+
+The Block Validator validates blocks except from the perspective of consensus.
+
+**Chain Controller**
+
+The Chain Controller manages both committed and uncommitted chain heads. The
+Chain Controller is driven by the Consensus Engine and does not know about
+block validity or consensus.
+
+**Block Publisher**
+
+The Block Publisher handles the creation of blocks. It is directed by the
+Consensus Engine on when to do this. When a block is finalized, it is forwarded
+to the Block Validator.
+
```

Comment:
Block Store and Block Cache from the diagram are not defined here. No need for unnecessary work if the definitions remain the same as here: https://sawtooth.hyperledger.org/docs/core/releases/latest/architecture/journal.html.